pkg/resource/deploy/source_eval.go

```diff
@@ -784,8 +825,19 @@ func (rm *resmon) Address() string {
 
 // Cancel signals that the engine should be terminated, awaits its termination, and returns any errors that result.
 func (rm *resmon) Cancel() error {
+	// By closing `rm.cancel` we cancel all in flight steps and initiate the
+	// graceful shutdown of the server. We won't accept any new connections, but
+	// pending connections will be allowed to complete.
+	//
+	// We need to do this before receiving from `rm.programCompleteChan`,
+	// otherwise we may deadlock: For example a `RegisterResource` call will not
+	// return to the program until we cancel, and we will never write to
+	// `rm.programCompleteChan` unless the program exits.
 	close(rm.cancel)
-	errs := []error{<-rm.done}
+	var err error
+	rm.waitForShutdownChan <- struct{}{} // Signal to the program that we are ready to shutdown ...
+	err = <-rm.programCompleteChan       // and wait for it to complete.
```

Comment:
I don't think we need to wait for this, we can just wait for the program to finish like normal